driver/XDMA/linux-kernel/xdma/tsn.c

```diff
@@ -423,30 +416,22 @@ static bool get_timestamps(struct timestamps* timestamps, const struct tsn_confi
 	return true;
 }
 
-static bool append_buffer_track(struct buffer_tracker* buffer_tracker) {
-	if (buffer_tracker->pending_packets >= HW_QUEUE_SIZE) {
-		return false;
+static void tsn_buffer_tracker_update(struct xdma_dev* xdev) {
+	if (xdev->tsn_config.buffer_tracker.available_space < HW_QUEUE_SIZE_PAD) {
+		u64 total_pkts = alinx_get_total_new_entry_by_xdev(xdev);
+		u64 sent_pkts = alinx_get_total_valid_entry_by_xdev(xdev);
+		u64 dropped_pkts = alinx_get_total_drop_entry_by_xdev(xdev);
+		xdev->tsn_config.buffer_tracker.available_space = HW_QUEUE_SIZE - (u32)(total_pkts - sent_pkts - dropped_pkts);
 	}
-
-	buffer_tracker->pending_packets += 1;
-	return true;
 }
 
-static void update_buffer_track(struct pci_dev* pdev) {
-	struct xdma_dev* xdev = xdev_find_by_pdev(pdev);
-	struct buffer_tracker* buffer_tracker = &xdev->tsn_config.buffer_tracker;
-	u64 tx_count, pop_count;
-
-	if (buffer_tracker->pending_packets < HW_QUEUE_SIZE - HW_QUEUE_SIZE_PAD) {
-		// No need to update that frequently
-		return;
-	}
+static void tsn_buffer_tracker_add_packet(struct xdma_dev* xdev) {
```

Comment:
```suggestion
static void decrease_available_tsn_buffer_space(struct xdma_dev* xdev) {
```

tsn_buffer_tracker를 함수명의 앞에 두어야하는 규칙이 정해져있는 것 일수도 있으니, 만약 그러하다면 뒤에 붙는 add_packet 말고 **감소**의 느낌을 주는 영단어로 수정부탁드립니다.

Reply:
또한, tsn_config 하위의 buffer_tracker는 TSN 기능이랑 연관이 있는 걸까요?

Reply:
아뇨, 오로지 TSN NIC의 버퍼 상태만 추적하기 위한 구조체입니다.

1Q TSN 로직 업데이트 전에는 굉장히 복잡한 방식으로 버퍼 상태를 추적했어서,
(버퍼에서 빠진 프레임의 수가 그 이유에 따라 여기저기 흩어지고, 일부 레지스터는 읽으면 초기화되지만 일부는 초기화되지 않는 등)
관리하기 편하도록 트래커를 따로 만들었던 것입니다.

지금은 굳이 그럴 필요가 없어져서, buffer_tracker는 없애고 available_space 변수 하나만 남겨도 될 것 같습니다.

Reply:
buffer_tracker를 삭제하고 buffer_space 변수로 대체했습니다.